chapter02/worksheets/solution2.2.6.3.scala

```diff
@@ -0,0 +1,25 @@
+/** 
+  Exercise 2.2.6.3
+  Use foldLeft to implement the zipWithIndex method for sequences.
+
+  The required type signature and a sample test:
+
+  def zipWithIndex[A](xs: Seq[A]): Seq[(A, Int)] = ???
+
+  scala> zipWithIndex(Seq("a", "b", "c", "d"))
+  res0: Seq[(String, Int)] = List((a, 0), (b, 1), (c, 2), (d, 3))
+  */
+
+def zipWithIndex[A](xs: Seq[A]): Seq[(A, Int)] = {
+  val a =  (0 to xs.length - 1)
+  type Acc = Seq[(A, Int)]
+  var counter = (-1)
```

Comment:
The solution should not use `var`.

I suggest that `type Acc` should include the current value of the index as well as the resulting sequence.

Reply:
You wrote a loop that works. The loop goes like this:
- Initially, counter = -1 and the sequence is empty.
- At each step, counter = counter + 1 and the sequence is modified as well.

To reformulate this as a `fold`, you try writing code like this:

```scala
def zipWithIndex[A](xs: Seq[A]): Seq[(A, Int)] = {
   type Acc = ???
   val init : Acc = ???
   val update: (Acc, A) => Acc = { case (previousAcc, x) =>  ??? }

   xs.foldLeft(init)(update)
}
```

The type `Acc` must hold all data that is being updated by the loop. In your case, this is both the counter and the sequence.
So, the accumulator must be a tuple: `Acc = (Int, Seq[A])`.
Now you can rewrite your loop as a `foldLeft`. No more `var` anywhere.
You just need to write `init` and `update`.

The function `update` will look like this: `{ case ((previousCounter, previousSeq), x) =>  (newCounter, newSeq) }` where you need to write the correct expressions for `newCounter` and `newSeq`.